internal/loadbalancer/balance_chwbl.go

```diff
@@ -0,0 +1,142 @@
+package loadbalancer
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/cespare/xxhash"
+)
+
+func (g *group) chwblGetAddr(key string, loadFactor float64, adapter string) (endpoint, bool) {
+	if len(g.chwblHashes) == 0 {
+		return endpoint{}, false
+	}
+
+	h := chwblHash(key)
+	_, i0 := g.chwblSearch(h)
+
+	var defaultEndpoint *endpoint
+
+	i := i0
+	// Avoid an infinite loop by checking if we've checked all the endpoints.
+	for n := 0; n < len(g.chwblSortedHashes); n++ {
+		name := g.chwblHashes[g.chwblSortedHashes[i]]
+		ep, ok := g.endpoints[name]
+		if !ok {
+			panic(fmt.Sprintf("endpoints corrupted, %q should be in map", name))
+		}
+
+		var adapterMatches bool
+		if adapter == "" {
+			adapterMatches = true
+		} else {
+			_, adapterMatches = ep.adapters[adapter]
+		}
+
+		if adapterMatches {
+			if defaultEndpoint == nil {
+				// Save the first endpoint that has the adapter in case no
+				// endpoint is found with acceptable load.
+				defaultEndpoint = &ep
+			}
+			if chwblLoadOK(ep.inFlight.Load(), g.totalInFlight.Load(), len(g.endpoints), loadFactor) {
```

Comment:
nit: this can be done before setting the default endpoint.

Reply:
The default endpoint is the endpoint that is able to serve the request (has the adapter) but might not meet the load requirement after all other endpoints have been checked.

Adding comment.